---
title: "RNA-seq differential exp"
output: html_document
date: "2023-05-23"
---

```{r include=FALSE}
library(plyr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(edgeR)
library(reshape2)
```


## Data pre-treatment

* files downloaded from TCGA = 429
  * stage_I = 1
  * stage_II = 26
  * stage_III = 335
  * stage_IV = 64
  
* sort_tcga_files.R: 
  * Uses clinical and gdc_sample_sheet to sort files downloaded files by stage
  * Moves them into appropriate directories

* tcga/kylie_master_df.R: 
  * combines all the files within each stage into a single dataframe 
  * subsets unstranded counts

<br>

<br>

## TCGA data edgeR pipeline {.tabset .tabset-fade .tabset-pills}

### Stage_I
```{r echo=FALSE}
stage <- "stage_I"

gtex_data <- read.table("gtex/gene_reads_2017-06-05_v8_ovary.gct", skip = 2)

# set col names of gtex data
colnames(gtex_data) <- gtex_data[1, ]
gtex_data <- gtex_data[-1, ]

# rename cols
colnames(gtex_data)[2] <- "gene_id"
colnames(gtex_data)[3] <- "gene_name"
gtex_data <- gtex_data[, -1]


tcga_data <- read.csv(paste0("tcga/", stage, "_master_df.csv"), row.names = 1)

# remove number after dot point in ensembl id in both files
gtex_data$gene_id <- gsub("\\.[0-9]*$", "", gtex_data$gene_id)
tcga_data$gene_id <- gsub("\\.[0-9]*$", "", tcga_data$gene_id)

# get rid of gene_name columns
tcga_data <- tcga_data[, -2]
gtex_data <- gtex_data[, -2]

df <- inner_join(tcga_data, gtex_data, by = c("gene_id"))
rownames(df) <- df$gene_id
df <- df[, -1]
# convert to numeric because for some reason they are as chr
df[] <- apply(df, 2, as.numeric)

# subset genes that have a cpm of greater than one in more than 50% of the samples
keepTheseGenes <- (rowSums(cpm(df) > 1) >= ncol(df)/2)
print("Kept genes")
print(summary(keepTheseGenes))

df <- df[keepTheseGenes, ]

# Select the columns that start with "unstranded"
cols <- grep("^unstranded", colnames(df))

# Assign the unstranded columns to the cancer group
cancer <- colnames(df)[cols]

# Assign the rest of the columns to the healthy group
healthy <- setdiff(colnames(df), cancer)

# Create the group variable
group <- factor(c(rep("cancer", length(cancer)), rep("healthy", length(healthy))))

data <- DGEList(counts = df, group = group)

design <- model.matrix(~group)

# Estimate a common negative binomial dispersion parameter for a DGE dataset with a general experimental design
common <- estimateGLMCommonDisp(data, design, verbose = T)

# Estimate the abundance-dispersion trend by Cox-Reid approximate profile likelihood.
trend <- estimateGLMTrendedDisp(common, design)

# Compute an empirical Bayes estimate of the negative binomial dispersion parameter for each tag, 
# with expression levels specified by a log-linear model.
tagwise <- estimateGLMTagwiseDisp(trend, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
fit <- glmFit(tagwise, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
lrt <- glmLRT(fit, coef = 2)

# Extract the most differentially expressed genes (or sequence tags) from a test object, 
# ranked either by p-value or by absolute log-fold-change.
toptags <- topTags(lrt, n = Inf)

# Identify which genes are significantly differentially expressed from 
# an edgeR fit object containing p-values and test statistics.
dif_exp <- decideTestsDGE(lrt, p = 0.05, adjust = "fdr", lfc = 1)
print(summary(dif_exp))

dif_exp_genes <- rownames(tagwise)[as.logical(dif_exp)]

# create a results df
hits <- toptags$table[toptags$table$FDR < 0.1, ]
colnames <- colnames(hits)
hits$gene_id <- rownames(hits)
hits <- hits[,c("gene_id", colnames)]
```


```{r echo=FALSE, out.width='33%', fig.show='hold'}
# Plot the genewise biological coefficient of variation (BCV) against gene abundance (in log2 counts per million).
plotBCV(tagwise)

# Make a mean-difference plot of two libraries of count data with smearing of points with very low counts, 
# especially those that are zero for one of the columns.
plotSmear(lrt, de.tags = dif_exp_genes)

# plot Pvalues of different logFC scores
ggplot(hits, aes(x=logFC, y=PValue)) + geom_point(size = 0.5) + labs(title = stage)
```

### Stage_II
```{r echo=FALSE}
stage <- "stage_II"

gtex_data <- read.table("gtex/gene_reads_2017-06-05_v8_ovary.gct", skip = 2)

# set col names of gtex data
colnames(gtex_data) <- gtex_data[1, ]
gtex_data <- gtex_data[-1, ]

# rename cols
colnames(gtex_data)[2] <- "gene_id"
colnames(gtex_data)[3] <- "gene_name"
gtex_data <- gtex_data[, -1]


tcga_data <- read.csv(paste0("tcga/", stage, "_master_df.csv"), row.names = 1)

# remove number after dot point in ensembl id in both files
gtex_data$gene_id <- gsub("\\.[0-9]*$", "", gtex_data$gene_id)
tcga_data$gene_id <- gsub("\\.[0-9]*$", "", tcga_data$gene_id)

# get rid of gene_name columns
tcga_data <- tcga_data[, -2]
gtex_data <- gtex_data[, -2]

df <- inner_join(tcga_data, gtex_data, by = c("gene_id"))
rownames(df) <- df$gene_id
df <- df[, -1]
# convert to numeric because for some reason they are as chr
df[] <- apply(df, 2, as.numeric)

# subset genes that have a cpm of greater than one in more than 50% of the samples
keepTheseGenes <- (rowSums(cpm(df) > 1) >= ncol(df)/2)
print("Kept genes")
print(summary(keepTheseGenes))

df <- df[keepTheseGenes, ]

# Select the columns that start with "unstranded"
cols <- grep("^unstranded", colnames(df))

# Assign the unstranded columns to the cancer group
cancer <- colnames(df)[cols]

# Assign the rest of the columns to the healty group
healthy <- setdiff(colnames(df), cancer)

# Create the group variable
group <- factor(c(rep("cancer", length(cancer)), rep("healthy", length(healthy))))

data <- DGEList(counts = df, group = group)

design <- model.matrix(~group)

# Estimate a common negative binomial dispersion parameter for a DGE dataset with a general experimental design
common <- estimateGLMCommonDisp(data, design, verbose = T)

# Estimate the abundance-dispersion trend by Cox-Reid approximate profile likelihood.
trend <- estimateGLMTrendedDisp(common, design)

# Compute an empirical Bayes estimate of the negative binomial dispersion parameter for each tag, 
# with expression levels specified by a log-linear model.
tagwise <- estimateGLMTagwiseDisp(trend, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
fit <- glmFit(tagwise, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
lrt <- glmLRT(fit, coef = 2)

# Extract the most differentially expressed genes (or sequence tags) from a test object, 
# ranked either by p-value or by absolute log-fold-change.
toptags <- topTags(lrt, n = Inf)

# Identify which genes are significantly differentially expressed from 
# an edgeR fit object containing p-values and test statistics.
dif_exp <- decideTestsDGE(lrt, p = 0.05, adjust = "fdr", lfc = 1)
print(summary(dif_exp))

dif_exp_genes <- rownames(tagwise)[as.logical(dif_exp)]

# create a results df
hits <- toptags$table[toptags$table$FDR < 0.1, ]
colnames <- colnames(hits)
hits$gene_id <- rownames(hits)
hits <- hits[,c("gene_id", colnames)]
```


```{r echo=FALSE, out.width='33%', fig.show='hold'}
# Plot the genewise biological coefficient of variation (BCV) against gene abundance (in log2 counts per million).
plotBCV(tagwise)

# Make a mean-difference plot of two libraries of count data with smearing of points with very low counts, 
# especially those that are zero for one of the columns.
plotSmear(lrt, de.tags = dif_exp_genes)

# plot Pvalues of different logFC scores
ggplot(hits, aes(x=logFC, y=PValue)) + geom_point(size = 0.5) + labs(title = stage)
```

### Stage_III
```{r echo=FALSE}
stage <- "stage_III"

gtex_data <- read.table("gtex/gene_reads_2017-06-05_v8_ovary.gct", skip = 2)

# set col names of gtex data
colnames(gtex_data) <- gtex_data[1, ]
gtex_data <- gtex_data[-1, ]

# rename cols
colnames(gtex_data)[2] <- "gene_id"
colnames(gtex_data)[3] <- "gene_name"
gtex_data <- gtex_data[, -1]


tcga_data <- read.csv(paste0("tcga/", stage, "_master_df.csv"), row.names = 1)

# remove number after dot point in ensembl id in both files
gtex_data$gene_id <- gsub("\\.[0-9]*$", "", gtex_data$gene_id)
tcga_data$gene_id <- gsub("\\.[0-9]*$", "", tcga_data$gene_id)

# get rid of gene_name columns
tcga_data <- tcga_data[, -2]
gtex_data <- gtex_data[, -2]

df <- inner_join(tcga_data, gtex_data, by = c("gene_id"))
rownames(df) <- df$gene_id
df <- df[, -1]
# convert to numeric because for some reason they are as chr
df[] <- apply(df, 2, as.numeric)

# subset genes that have a cpm of greater than one in more than 50% of the samples
keepTheseGenes <- (rowSums(cpm(df) > 1) >= ncol(df)/2)
print("Kept genes")
print(summary(keepTheseGenes))

df <- df[keepTheseGenes, ]

# Select the columns that start with "unstranded"
cols <- grep("^unstranded", colnames(df))

# Assign the unstranded columns to the cancer group
cancer <- colnames(df)[cols]

# Assign the rest of the columns to the healty group
healthy <- setdiff(colnames(df), cancer)

# Create the group variable
group <- factor(c(rep("cancer", length(cancer)), rep("healthy", length(healthy))))

data <- DGEList(counts = df, group = group)

design <- model.matrix(~group)

# Estimate a common negative binomial dispersion parameter for a DGE dataset with a general experimental design
common <- estimateGLMCommonDisp(data, design, verbose = T)

# Estimate the abundance-dispersion trend by Cox-Reid approximate profile likelihood.
trend <- estimateGLMTrendedDisp(common, design)

# Compute an empirical Bayes estimate of the negative binomial dispersion parameter for each tag, 
# with expression levels specified by a log-linear model.
tagwise <- estimateGLMTagwiseDisp(trend, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
fit <- glmFit(tagwise, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
lrt <- glmLRT(fit, coef = 2)

# Extract the most differentially expressed genes (or sequence tags) from a test object, 
# ranked either by p-value or by absolute log-fold-change.
toptags <- topTags(lrt, n = Inf)

# Identify which genes are significantly differentially expressed from 
# an edgeR fit object containing p-values and test statistics.
dif_exp <- decideTestsDGE(lrt, p = 0.05, adjust = "fdr", lfc = 1)
print(summary(dif_exp))

dif_exp_genes <- rownames(tagwise)[as.logical(dif_exp)]

# create a results df
hits <- toptags$table[toptags$table$FDR < 0.1, ]
colnames <- colnames(hits)
hits$gene_id <- rownames(hits)
hits <- hits[,c("gene_id", colnames)]
```


```{r echo=FALSE, out.width='33%', fig.show='hold'}
# Plot the genewise biological coefficient of variation (BCV) against gene abundance (in log2 counts per million).
plotBCV(tagwise)

# Make a mean-difference plot of two libraries of count data with smearing of points with very low counts, 
# especially those that are zero for one of the columns.
plotSmear(lrt, de.tags = dif_exp_genes)

# plot Pvalues of different logFC scores
ggplot(hits, aes(x=logFC, y=PValue)) + geom_point(size = 0.5) + labs(title = stage)
```

### Stage_IV
```{r echo=FALSE}
stage <- "stage_IV"

gtex_data <- read.table("gtex/gene_reads_2017-06-05_v8_ovary.gct", skip = 2)

# set col names of gtex data
colnames(gtex_data) <- gtex_data[1, ]
gtex_data <- gtex_data[-1, ]

# rename cols
colnames(gtex_data)[2] <- "gene_id"
colnames(gtex_data)[3] <- "gene_name"
gtex_data <- gtex_data[, -1]


tcga_data <- read.csv(paste0("tcga/", stage, "_master_df.csv"), row.names = 1)

# remove number after dot point in ensembl id in both files
gtex_data$gene_id <- gsub("\\.[0-9]*$", "", gtex_data$gene_id)
tcga_data$gene_id <- gsub("\\.[0-9]*$", "", tcga_data$gene_id)

# get rid of gene_name columns
tcga_data <- tcga_data[, -2]
gtex_data <- gtex_data[, -2]

df <- inner_join(tcga_data, gtex_data, by = c("gene_id"))
rownames(df) <- df$gene_id
df <- df[, -1]
# convert to numeric because for some reason they are as chr
df[] <- apply(df, 2, as.numeric)

# subset genes that have a cpm of greater than one in more than 50% of the samples
keepTheseGenes <- (rowSums(cpm(df) > 1) >= ncol(df)/2)
print("Kept genes")
print(summary(keepTheseGenes))

df <- df[keepTheseGenes, ]

# Select the columns that start with "unstranded"
cols <- grep("^unstranded", colnames(df))

# Assign the unstranded columns to the cancer group
cancer <- colnames(df)[cols]

# Assign the rest of the columns to the healty group
healthy <- setdiff(colnames(df), cancer)

# Create the group variable
group <- factor(c(rep("cancer", length(cancer)), rep("healthy", length(healthy))))

data <- DGEList(counts = df, group = group)

design <- model.matrix(~group)

# Estimate a common negative binomial dispersion parameter for a DGE dataset with a general experimental design
common <- estimateGLMCommonDisp(data, design, verbose = T)

# Estimate the abundance-dispersion trend by Cox-Reid approximate profile likelihood.
trend <- estimateGLMTrendedDisp(common, design)

# Compute an empirical Bayes estimate of the negative binomial dispersion parameter for each tag, 
# with expression levels specified by a log-linear model.
tagwise <- estimateGLMTagwiseDisp(trend, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
fit <- glmFit(tagwise, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
lrt <- glmLRT(fit, coef = 2)

# Extract the most differentially expressed genes (or sequence tags) from a test object, 
# ranked either by p-value or by absolute log-fold-change.
toptags <- topTags(lrt, n = Inf)

# Identify which genes are significantly differentially expressed from 
# an edgeR fit object containing p-values and test statistics.
dif_exp <- decideTestsDGE(lrt, p = 0.05, adjust = "fdr", lfc = 1)
print(summary(dif_exp))

dif_exp_genes <- rownames(tagwise)[as.logical(dif_exp)]

# create a results df
hits <- toptags$table[toptags$table$FDR < 0.1, ]
colnames <- colnames(hits)
hits$gene_id <- rownames(hits)
hits <- hits[,c("gene_id", colnames)]
```


```{r echo=FALSE, out.width='33%', fig.show='hold'}
# Plot the genewise biological coefficient of variation (BCV) against gene abundance (in log2 counts per million).
plotBCV(tagwise)

# Make a mean-difference plot of two libraries of count data with smearing of points with very low counts, 
# especially those that are zero for one of the columns.
plotSmear(lrt, de.tags = dif_exp_genes)

# plot Pvalues of different logFC scores
ggplot(hits, aes(x=logFC, y=PValue)) + geom_point(size = 0.5) + labs(title = stage)
```

<br>

<br>

## Kylie's data edgeR pipeline {.tabset .tabset-fade .tabset-pills}

### Stage_I
```{r}
stage <- "stage_I"

gtex_data <- read.table("gtex/gene_reads_2017-06-05_v8_ovary.gct", skip = 2)

# set col names of gtex data
colnames(gtex_data) <- gtex_data[1, ]
gtex_data <- gtex_data[-1, ]

# rename cols
colnames(gtex_data)[2] <- "gene_id"
colnames(gtex_data)[3] <- "gene_name"
gtex_data <- gtex_data[, -1]

# remove number after dot point in ensembl id
gtex_data$gene_id <- gsub("\\.[0-9]*$", "", gtex_data$gene_id)

# get rid of gene name column
gtex_data <- gtex_data[, -2]

kylie_data <- read.csv(paste0("kylie/", stage, "_master_df.csv"))
colnames(kylie_data)[1] <- "gene_id"

# combine into single df
df <- inner_join(kylie_data, gtex_data, by = c("gene_id"))
rownames(df) <- df$gene_id
df <- df[, -1]
# convert to numeric because for some reason they are as chr
df[] <- apply(df, 2, as.numeric)

# subset genes that have a cpm of greater than one in more than 50% of the samples
keepTheseGenes <- (rowSums(cpm(df) > 1) >= ncol(df)/2)
print("Kept genes")
print(summary(keepTheseGenes))

df <- df[keepTheseGenes, ]

# Select the columns that start with "unstranded"
cols <- grep("^GAMuT", colnames(df))

# Assign the unstranded columns to the cancer group
cancer <- colnames(df)[cols]

# Assign the rest of the columns to the healty group
healthy <- setdiff(colnames(df), cancer)

# Create the group variable
group <- factor(c(rep("cancer", length(cancer)), rep("healthy", length(healthy))))

data <- DGEList(counts = df, group = group)

design <- model.matrix(~group)

# Estimate a common negative binomial dispersion parameter for a DGE dataset with a general experimental design
common <- estimateGLMCommonDisp(data, design, verbose = T)

# Estimate the abundance-dispersion trend by Cox-Reid approximate profile likelihood.
trend <- estimateGLMTrendedDisp(common, design)

# Compute an empirical Bayes estimate of the negative binomial dispersion parameter for each tag, 
# with expression levels specified by a log-linear model.
tagwise <- estimateGLMTagwiseDisp(trend, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
fit <- glmFit(tagwise, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
lrt <- glmLRT(fit, coef = 2)

# Extract the most differentially expressed genes (or sequence tags) from a test object, 
# ranked either by p-value or by absolute log-fold-change.
toptags <- topTags(lrt, n = Inf)

# Identify which genes are significantly differentially expressed from 
# an edgeR fit object containing p-values and test statistics.
dif_exp <- decideTestsDGE(lrt, p = 0.05, adjust = "fdr", lfc = 1)
print(summary(dif_exp))

dif_exp_genes <- rownames(tagwise)[as.logical(dif_exp)]

# create a results df
hits <- toptags$table[toptags$table$FDR < 0.1, ]
colnames <- colnames(hits)
hits$gene_id <- rownames(hits)
hits <- hits[,c("gene_id", colnames)]
```

```{r echo=FALSE, out.width='33%', fig.show='hold'}
# Plot the genewise biological coefficient of variation (BCV) against gene abundance (in log2 counts per million).
plotBCV(tagwise)

# Make a mean-difference plot of two libraries of count data with smearing of points with very low counts, 
# especially those that are zero for one of the columns.
plotSmear(lrt, de.tags = dif_exp_genes)

# plot Pvalues of different logFC scores
ggplot(hits, aes(x=logFC, y=PValue)) + geom_point(size = 0.5) + labs(title = stage)
```

### Stage_II
```{r}
stage <- "stage_II"

gtex_data <- read.table("gtex/gene_reads_2017-06-05_v8_ovary.gct", skip = 2)

# set col names of gtex data
colnames(gtex_data) <- gtex_data[1, ]
gtex_data <- gtex_data[-1, ]

# rename cols
colnames(gtex_data)[2] <- "gene_id"
colnames(gtex_data)[3] <- "gene_name"
gtex_data <- gtex_data[, -1]

# remove number after dot point in ensembl id
gtex_data$gene_id <- gsub("\\.[0-9]*$", "", gtex_data$gene_id)

# get rid of gene name column
gtex_data <- gtex_data[, -2]

kylie_data <- read.csv(paste0("kylie/", stage, "_master_df.csv"))
colnames(kylie_data)[1] <- "gene_id"

# combine into single df
df <- inner_join(kylie_data, gtex_data, by = c("gene_id"))
rownames(df) <- df$gene_id
df <- df[, -1]
# convert to numeric because for some reason they are as chr
df[] <- apply(df, 2, as.numeric)

# subset genes that have a cpm of greater than one in more than 50% of the samples
keepTheseGenes <- (rowSums(cpm(df) > 1) >= ncol(df)/2)
print("Kept genes")
print(summary(keepTheseGenes))

df <- df[keepTheseGenes, ]

# Select the columns that start with "unstranded"
cols <- grep("^GAMuT", colnames(df))

# Assign the unstranded columns to the cancer group
cancer <- colnames(df)[cols]

# Assign the rest of the columns to the healty group
healthy <- setdiff(colnames(df), cancer)

# Create the group variable
group <- factor(c(rep("cancer", length(cancer)), rep("healthy", length(healthy))))

data <- DGEList(counts = df, group = group)

design <- model.matrix(~group)

# Estimate a common negative binomial dispersion parameter for a DGE dataset with a general experimental design
common <- estimateGLMCommonDisp(data, design, verbose = T)

# Estimate the abundance-dispersion trend by Cox-Reid approximate profile likelihood.
trend <- estimateGLMTrendedDisp(common, design)

# Compute an empirical Bayes estimate of the negative binomial dispersion parameter for each tag, 
# with expression levels specified by a log-linear model.
tagwise <- estimateGLMTagwiseDisp(trend, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
fit <- glmFit(tagwise, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
lrt <- glmLRT(fit, coef = 2)

# Extract the most differentially expressed genes (or sequence tags) from a test object, 
# ranked either by p-value or by absolute log-fold-change.
toptags <- topTags(lrt, n = Inf)

# Identify which genes are significantly differentially expressed from 
# an edgeR fit object containing p-values and test statistics.
dif_exp <- decideTestsDGE(lrt, p = 0.05, adjust = "fdr", lfc = 1)
print(summary(dif_exp))

dif_exp_genes <- rownames(tagwise)[as.logical(dif_exp)]

# create a results df
hits <- toptags$table[toptags$table$FDR < 0.1, ]
colnames <- colnames(hits)
hits$gene_id <- rownames(hits)
hits <- hits[,c("gene_id", colnames)]
```

```{r echo=FALSE, out.width='33%', fig.show='hold'}
# Plot the genewise biological coefficient of variation (BCV) against gene abundance (in log2 counts per million).
plotBCV(tagwise)

# Make a mean-difference plot of two libraries of count data with smearing of points with very low counts, 
# especially those that are zero for one of the columns.
plotSmear(lrt, de.tags = dif_exp_genes)

# plot Pvalues of different logFC scores
ggplot(hits, aes(x=logFC, y=PValue)) + geom_point(size = 0.5) + labs(title = stage)
```

### Stage_III
```{r}
stage <- "stage_III"

gtex_data <- read.table("gtex/gene_reads_2017-06-05_v8_ovary.gct", skip = 2)

# set col names of gtex data
colnames(gtex_data) <- gtex_data[1, ]
gtex_data <- gtex_data[-1, ]

# rename cols
colnames(gtex_data)[2] <- "gene_id"
colnames(gtex_data)[3] <- "gene_name"
gtex_data <- gtex_data[, -1]

# remove number after dot point in ensembl id
gtex_data$gene_id <- gsub("\\.[0-9]*$", "", gtex_data$gene_id)

# get rid of gene name column
gtex_data <- gtex_data[, -2]

kylie_data <- read.csv(paste0("kylie/", stage, "_master_df.csv"))
colnames(kylie_data)[1] <- "gene_id"

# combine into single df
df <- inner_join(kylie_data, gtex_data, by = c("gene_id"))
rownames(df) <- df$gene_id
df <- df[, -1]
# convert to numeric because for some reason they are as chr
df[] <- apply(df, 2, as.numeric)

# subset genes that have a cpm of greater than one in more than 50% of the samples
keepTheseGenes <- (rowSums(cpm(df) > 1) >= ncol(df)/2)
print("Kept genes")
print(summary(keepTheseGenes))

df <- df[keepTheseGenes, ]

# Select the columns that start with "unstranded"
cols <- grep("^GAMuT", colnames(df))

# Assign the unstranded columns to the cancer group
cancer <- colnames(df)[cols]

# Assign the rest of the columns to the healty group
healthy <- setdiff(colnames(df), cancer)

# Create the group variable
group <- factor(c(rep("cancer", length(cancer)), rep("healthy", length(healthy))))

data <- DGEList(counts = df, group = group)

design <- model.matrix(~group)

# Estimate a common negative binomial dispersion parameter for a DGE dataset with a general experimental design
common <- estimateGLMCommonDisp(data, design, verbose = T)

# Estimate the abundance-dispersion trend by Cox-Reid approximate profile likelihood.
trend <- estimateGLMTrendedDisp(common, design)

# Compute an empirical Bayes estimate of the negative binomial dispersion parameter for each tag, 
# with expression levels specified by a log-linear model.
tagwise <- estimateGLMTagwiseDisp(trend, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
fit <- glmFit(tagwise, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
lrt <- glmLRT(fit, coef = 2)

# Extract the most differentially expressed genes (or sequence tags) from a test object, 
# ranked either by p-value or by absolute log-fold-change.
toptags <- topTags(lrt, n = Inf)

# Identify which genes are significantly differentially expressed from 
# an edgeR fit object containing p-values and test statistics.
dif_exp <- decideTestsDGE(lrt, p = 0.05, adjust = "fdr", lfc = 1)
print(summary(dif_exp))

dif_exp_genes <- rownames(tagwise)[as.logical(dif_exp)]

# create a results df
hits <- toptags$table[toptags$table$FDR < 0.1, ]
colnames <- colnames(hits)
hits$gene_id <- rownames(hits)
hits <- hits[,c("gene_id", colnames)]
```

```{r echo=FALSE, out.width='33%', fig.show='hold'}
# Plot the genewise biological coefficient of variation (BCV) against gene abundance (in log2 counts per million).
plotBCV(tagwise)

# Make a mean-difference plot of two libraries of count data with smearing of points with very low counts, 
# especially those that are zero for one of the columns.
plotSmear(lrt, de.tags = dif_exp_genes)

# plot Pvalues of different logFC scores
ggplot(hits, aes(x=logFC, y=PValue)) + geom_point(size = 0.5) + labs(title = stage)
```

### Stage_IV
```{r}
stage <- "stage_IV"

gtex_data <- read.table("gtex/gene_reads_2017-06-05_v8_ovary.gct", skip = 2)

# set col names of gtex data
colnames(gtex_data) <- gtex_data[1, ]
gtex_data <- gtex_data[-1, ]

# rename cols
colnames(gtex_data)[2] <- "gene_id"
colnames(gtex_data)[3] <- "gene_name"
gtex_data <- gtex_data[, -1]

# remove number after dot point in ensembl id
gtex_data$gene_id <- gsub("\\.[0-9]*$", "", gtex_data$gene_id)

# get rid of gene name column
gtex_data <- gtex_data[, -2]

kylie_data <- read.csv(paste0("kylie/", stage, "_master_df.csv"))
colnames(kylie_data)[1] <- "gene_id"

# combine into single df
df <- inner_join(kylie_data, gtex_data, by = c("gene_id"))
rownames(df) <- df$gene_id
df <- df[, -1]
# convert to numeric because for some reason they are as chr
df[] <- apply(df, 2, as.numeric)

# subset genes that have a cpm of greater than one in more than 50% of the samples
keepTheseGenes <- (rowSums(cpm(df) > 1) >= ncol(df)/2)
print("Kept genes")
print(summary(keepTheseGenes))

df <- df[keepTheseGenes, ]

# Select the columns that start with "unstranded"
cols <- grep("^GAMuT", colnames(df))

# Assign the unstranded columns to the cancer group
cancer <- colnames(df)[cols]

# Assign the rest of the columns to the healty group
healthy <- setdiff(colnames(df), cancer)

# Create the group variable
group <- factor(c(rep("cancer", length(cancer)), rep("healthy", length(healthy))))

data <- DGEList(counts = df, group = group)

design <- model.matrix(~group)

# Estimate a common negative binomial dispersion parameter for a DGE dataset with a general experimental design
common <- estimateGLMCommonDisp(data, design, verbose = T)

# Estimate the abundance-dispersion trend by Cox-Reid approximate profile likelihood.
trend <- estimateGLMTrendedDisp(common, design)

# Compute an empirical Bayes estimate of the negative binomial dispersion parameter for each tag, 
# with expression levels specified by a log-linear model.
tagwise <- estimateGLMTagwiseDisp(trend, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
fit <- glmFit(tagwise, design)

# Fit a negative binomial generalized log-linear model to the read counts for each gene. 
# Conduct genewise statistical tests for a given coefficient or coefficient contrast.
lrt <- glmLRT(fit, coef = 2)

# Extract the most differentially expressed genes (or sequence tags) from a test object, 
# ranked either by p-value or by absolute log-fold-change.
toptags <- topTags(lrt, n = Inf)

# Identify which genes are significantly differentially expressed from 
# an edgeR fit object containing p-values and test statistics.
dif_exp <- decideTestsDGE(lrt, p = 0.05, adjust = "fdr", lfc = 1)
print(summary(dif_exp))

dif_exp_genes <- rownames(tagwise)[as.logical(dif_exp)]

# create a results df
hits <- toptags$table[toptags$table$FDR < 0.1, ]
colnames <- colnames(hits)
hits$gene_id <- rownames(hits)
hits <- hits[,c("gene_id", colnames)]
```

```{r echo=FALSE, out.width='33%', fig.show='hold'}
# Plot the genewise biological coefficient of variation (BCV) against gene abundance (in log2 counts per million).
plotBCV(tagwise)

# Make a mean-difference plot of two libraries of count data with smearing of points with very low counts, 
# especially those that are zero for one of the columns.
plotSmear(lrt, de.tags = dif_exp_genes)

# plot Pvalues of different logFC scores
ggplot(hits, aes(x=logFC, y=PValue)) + geom_point(size = 0.5) + labs(title = stage)
```

<br>

<br>

## PCSF file prep

* Uses logFC from edgeR output file
* Converts gene ensembl to uniprot ID
* Create list of uniprot IDs for Cytoscape/STRINGapp
* Formats Cytoscape/STRINGapp output 

* Side note: Working on integrating Cytoscape/STRINGapp into R code

* Cytoscape/STRING node count:
  * TCGA stage_I: 6255
  * kylie stage_I: 5808
  
  * TCGA stage_II: 5736
  * Kylie stage_II: 6269
  
  * TCGA stage_III: 6013
  * Kylie stage_III: 6385
  
  * TCGA stage_IV: 5823
  * Kylie stage_IV: 6591

<br>

<br>

## PCSF results







